light_controller/ir_toy/ir_toy.py

```diff
@@ -0,0 +1,352 @@
+    0x13 LED off (FW v07+)
+    0x23 Settings descriptor report (FW v20+)
+    0x24 Enable transmit byte count report (FW v20+)
+    0x25 Enable transmit notify on complete (FW v20+)
+    0x26 Enable transmit handshake (FW v20+)
+    0x30 IO write
+    0x31 IO direction
+    0x32 IO read
+    0x40 UART setup
+    0x41 UART close
+    0x42 UART write
+    """
+
+    def __init__(self, comPort):
+        self._connection = serial.Serial(comPort)
+        self.protocolVersion = ""
+        self.flush()
+        self.setSamplingMode()
+
+    def __del__(self):
+        self.close()
+
+    def close(self):
+        if self._connection.isOpen():
+            self._connection.close()
+
+    def flush(self):
+        """
+        Flush the serial connection
+        :return: None
+        """
+        self._connection.flushInput()
+        self._connection.flush()
+        self._connection.flushOutput()
+
+    def read(self, num):
+        """
+        Read `num` bytes from the device
+        :param num: Number of bytes to read
+        :return: string of bytes
+        """
+        return self._connection.read(num)
+
+    def write(self, data):
+        """
+        Write data to the device
+        :param data: Data to be written
+        :return: Number of bytes written
+        """
+        result = self._connection.write(data)
+        return result
+
+    def writeArray(self, data):
+        """
+        Write a list of bytes
+        :param data: list of bytes
+        :return: number of bytes written
+        """
+        bytes = bytearray(data)
+        return self.write(bytes)
+
+    def writeByte(self, data):
+        """
+        Convert a byte to char and write it out
+        :param data: byte to write
+        :return: number of bytes written
+        """
+        return self.writeArray([data, ])
+
+    def enableTransmitMode(self):
+        """
+        Set transmission mode
+        :return: Size of TX buffer
+        """
+        self.writeByte(CommandEnum.transmit)
+        return ord(self.read(1))
+
+    def getFrequencyReport(self):
+        """
+        Get the raw timing information for the previous IR signal
+
+        :return: (timer at 2nd rising edge,
+                  timer at 3rd rising edge,
+                  timer at 4th rising edge,
+                  total count of pulses since last 0xFF)
+        """
+        self.writeByte(CommandEnum.getFrequencyReport)
+        buffer = self.read(8)
+        return struct.unpack(">4H", buffer)
+
+    def setSampleTimer(self, prescale):
+        """
+        The timer is 21.13333 us by default.
+        111 (0x07)	1:256	21.3333uS
+        110 (0x06)	1:128	10.6666uS
+        101 (0x05)	1:64
+        100 (0x04)	1:32
+        011 (0x03)	1:16
+        010 (0x02)	1:8
+        001 (0x01)	1:4
+        000 (0x00)	1:2
+        :param prescale: see above for valid values
+        :return: None
+        """
+        self.writeArray((CommandEnum.setSampleTimer, prescale))
```

Comment:
It's best to validate the value of `prescale` before we send this to the machine.